BuildScript/BuildTasks/MetadataValueDictionary.cs

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) rubicon IT GmbH, www.rubicon.eu
+//
+// See the NOTICE file distributed with this work for additional information
+// regarding copyright ownership.  rubicon licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may not use this
+// file except in compliance with the License.  You may obtain a copy of the
+// License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+// License for the specific language governing permissions and limitations
+// under the License.
+//
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Remotion.BuildScript.BuildTasks
+{
+  /// <summary>
+  /// Wraps a <see cref="IDictionary{TKey,TValue}"/> and performs case-insensitive access operations, emulating default msbuild behavior.
+  /// </summary>
+  public class MetadataValueDictionary
+  {
+    private readonly IReadOnlyDictionary<string, string> _rawDictionary;
+
+    public MetadataValueDictionary (IDictionary<string, string> rawDictionary)
+    {
+      _rawDictionary = new Dictionary<string, string> (rawDictionary, StringComparer.OrdinalIgnoreCase);
+    }
+
+    public KeyValuePair<string, string> Single (IEnumerable<string> possibleKeys, Func<Exception> createEmptySequenceException)
+    {
+      var enumeratedPossibleKeys = possibleKeys.ToArray();
+
+      var intersection = _rawDictionary.Keys.Intersect (enumeratedPossibleKeys, StringComparer.OrdinalIgnoreCase).ToArray();
+
+      if (intersection.Length > 1)
+      {
+        var jointPossibleKeys = string.Join (",", enumeratedPossibleKeys);
+        var jointActualKeys = string.Join (",", _rawDictionary.Keys);
+
+        throw new InvalidOperationException ($"Found multiple elements of '[{jointPossibleKeys}]' in '[{jointActualKeys}]'");
+      }
+
+      var key = intersection.SingleOrDefault() ?? throw createEmptySequenceException();
```

Comment:
This would perhaps feel more symetrical when written with 
IF > 1 throw
ELSE IF == 0 throw
ELSE return intersection[0]

Right now, we have to rely on branch coverage. The logic that SingleOrDefault is safe is only implicty established by reading the rest of the code.

Reply:
I have no issue with the change (even though intersection.Single() still seems more expressive to me), but sadly I can't understand your reasoning yet. Isn't `return intersection[0];` also only safe once one reads the rest of the code?

Reply:
The safety question isn't about to the use of Single / SingleOrDefault inside the branch but relies on the combination with the IF/ELIF/ELSE vs. using Single/SingleOrDefault as a standalone statement not connected to the guard condition.

There's only a few (< 10) occurances of the use of Single instead of a zero-index. So, what about Single() vs [0] in terms of idiomatic code? Usually, it feels more idiomatic when you see a zero-indexer in a guarded statement rather than Single. There's less semantic involved and it's more compact. If it's actually better or less readable once the other refactorings are done? Don't know, all I know is that I normally use the zero-indexer unless I try to be fancy with my code.